compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+          throw e
+        }
+      }
+
+    protected def compile(files0: Array[JFile], flags0: Array[String], suppressErrors: Boolean, targetDir: JFile): TestReporter = {
+
+      val flags = flags0 ++ Array("-d", targetDir.getAbsolutePath)
+
+      def flattenFiles(f: JFile): Array[JFile] =
+        if (f.isDirectory) f.listFiles.flatMap(flattenFiles)
+        else Array(f)
+
+      val files: Array[JFile] = files0.flatMap(flattenFiles)
+
+      def findJarFromRuntime(partialName: String) = {
+        val urls = ClassLoader.getSystemClassLoader.asInstanceOf[java.net.URLClassLoader].getURLs.map(_.getFile.toString)
+        urls.find(_.contains(partialName)).getOrElse {
+          throw new java.io.FileNotFoundException(
+            s"""Unable to locate $partialName on classpath:\n${urls.toList.mkString("\n")}"""
+          )
+        }
+      }
+
+      def addOutDir(xs: Array[String]): Array[String] = {
+        val (beforeCp, cpAndAfter) = xs.toList.span(_ != "-classpath")
+        if (cpAndAfter.nonEmpty) {
+          val (cp :: cpArg :: rest) = cpAndAfter
+          (beforeCp ++ (cp :: (cpArg + s":${targetDir.getAbsolutePath}") :: rest)).toArray
+        }
+        else (beforeCp ++ ("-classpath" :: targetDir.getAbsolutePath :: Nil)).toArray
+      }
+
+      def compileWithJavac(fs: Array[String]) = if (fs.nonEmpty) {
+        val scalaLib = findJarFromRuntime("scala-library-2.")
+        val fullArgs = Array(
+          "javac",
+          "-classpath",
+          s".:$scalaLib:${targetDir.getAbsolutePath}"
+        ) ++ flags.takeRight(2) ++ fs
+
+        Runtime.getRuntime.exec(fullArgs).waitFor() == 0
+      } else true
+
+
+      // First we try to compile the java files in the directory:
+      val javaFiles = files.filter(_.getName.endsWith(".java")).map(_.getAbsolutePath)
+      val javaCompiledBefore = compileWithJavac(javaFiles)
+
+      // Then we compile the scala files:
+      val reporter = TestReporter.parallelReporter(this, logLevel =
+        if (suppressErrors || suppressAllOutput) ERROR + 1 else ERROR)
+      val driver =
+        if (times == 1) new Driver { def newCompiler(implicit ctx: Context) = new Compiler }
+        else new Driver {
+          def newCompiler(implicit ctx: Context) = new Compiler
+
+          private def ntimes(n: Int)(op: Int => Reporter): Reporter =
+            (emptyReporter /: (1 to n)) ((_, i) => op(i))
+
+          override def doCompile(comp: Compiler, files: List[String])(implicit ctx: Context) =
+            ntimes(times) { run =>
+              val start = System.nanoTime()
+              val rep = super.doCompile(comp, files)
+              ctx.echo(s"\ntime run $run: ${(System.nanoTime - start) / 1000000}ms")
+              rep
+            }
+        }
+
+      val allArgs = addOutDir(flags)
+      driver.process(allArgs ++ files.map(_.getAbsolutePath), reporter = reporter)
+
+      // If the java files failed compilation before, we try again after:
```

Comment:
Imagine this:

```java
// Foo.java
pacakge foo;
public class Foo extends Bar {}
```

```scala
// Bar.scala
package foo
abstract class Bar
```

In this case, first compiling the java files and then the Scala files will fail. So we need to retry javac once if the initial compilation failed. If the reference is the other way around we first need to compile the Java files and then the Scala files anyway.